model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java

```diff
@@ -171,13 +175,29 @@ public Stream<E> read(QueryParameters<M> queryParameters) {
         }
         if (predicateFunc != null) query.where(predicateFunc.apply(cb, query::subquery, root));
 
+        // Add aliases for each root to avoid generating IDs with random object IDs in Hibernate 6.
+        // Those random object IDs would prevent creating a stable HQL string which is the key for caching.
+        // Use a HashMap to create increasing suffix numbers for each entity type.
+        Map<String, Integer> aliases = new HashMap<>();
+        for (Root<?> queryRoot : query.getRoots()) {
+            if (queryRoot.getAlias() == null) {
+                String name = queryRoot.getModel().getName();
+                Integer count = aliases.get(name);
+                if (count == null) {
+                    count = 0;
+                }
+                queryRoot.alias(name + count);
+                count ++;
+                aliases.put(name, count);
+            }
+        }
```

Comment:
Thank you @ahus1!

It seems to me that for each call of `read(QueryParameters<M> queryParameters)` there is empty map (`aliases`) created and for each root, which doesn't have an alias, it's created one. Then the alias is inserted into `aliases`.

Could you please provide an example of a query with multiple roots where alias would be found on line 185? So far I was not able to came across such a query. Thank you

Reply:
@vramik - I'm not aware of a secondary root being used at the moment. The closes I found was in JpaRoleModelCriteriaBuilder the subquery that's being created. 

I tried to trigger this code by calling a delete on  a role, but this this failed earlier as the "->" JSON functions weren't working (that's possibly another issue that we're about to soon for the H6 upgrade).

So to summarize: At the moment it is used only for one root, and I am not sure what it would do on the subquery. If the code would not set an alias for whatever reason, it would lead to a query that can't be cached, which might turn up in a performance test eventually.  

This whole caching code is currently used to cache the realm when working via the admin UI (and REST endpoints). It is also used to cache a client if a recall correctly during refresh of a token. I verified that those caching works as expected.

It still has its justification IMHO, at the same time we might skip it if we see it adds to much complexity.

Reply:
Makes sense, thanks. As far as I could see caching works as expected. My understating of the code above was that aliases are only stored into a map but there is never alias found in line 185. 

I just wanted to understand the purpose of the map properly so I wonder whether my understanding is correct or not.

Reply:
@vramik - yes, your understanding is correct. As there is currently only a single root, the logic isn't really tested used yet.  Today all aliases have the "0" at the end. For that to be "1", it would need to have a root with the same name a second time. 

I was assigning a separate numbering for each root as for two roots, I wouldn't know in which order I would receive them from the set of roots. So it was better to number those wich the same name consecutively instead of numbering all aliases consecutively.

Let me know if this comment helped to get this PR towards an approval for a review, or if you'd prefer to have a short call.

Reply:
Thanks for the comment, makes sense to me. Approving.